src/login/logind.c

```diff
@@ -409,10 +409,71 @@ static int manager_enumerate_users(Manager *m) {
         return r;
 }
 
+static int manager_attach_fds(Manager *m) {
+        _cleanup_strv_free_ char **fdnames = NULL;
+        int r = 0, n, i, fd;
+
+        /* Upon restart, PID1 will send us back all fds of session devices
+         * that we previously opened. Each file descriptor is associated
+         * with a given session. The session ids are passed through FDNAMES. */
+
+        n = sd_listen_fds_with_names(true, &fdnames);
+        if (n <= 0)
+                return n;
+
+        for (i = 0; i < n; i++) {
+                struct stat st;
+                SessionDevice *sd;
+                Session *s;
+                char *id;
+
+                fd = SD_LISTEN_FDS_START + i;
+
+                if (!startswith(fdnames[i], "session-"))
+                        continue;
+
+                id = fdnames[i] + 8;
+                s = hashmap_get(m->sessions, id);
+                if (!s) {
+                        /* If the session no more exists, the associated session
+                         * device to this fd doesn't either. Let's simply close this fd. */
+                        log_warning("Failed to attach fd for unknown session: %s", id);
+                        close_nointr(fd);
+                        r = -EINVAL;
+                        continue;
+                }
+
+                if (fstat(fd, &st) < 0) {
+                        log_warning("Failed to stat session device fd for session %s: %m", id);
```

Comment:
You cannot use %m without passing the errno code to log_warning. This needs to be `log_warning_errno(errno, ...)`.

Also, I don't think the error should be propagated. The device is allowed to go away at a random point, in which case `stat` failing is expected.